src/cargo/ops/cargo_compile.rs

```diff
@@ -155,7 +155,12 @@ impl<'a> Packages<'a> {
                     .collect()
             }
         };
-        Ok(specs)
+        if specs.is_empty() {
+            bail!("Workspace contains no members to be compiled. \
+                   Be sure all workspace members haven't been excluded")
```

Comment:
So, at this place I think we can give a more targeted error message, something like

```
let message = match self {
   Package::All => bail!("manifest path `{}` contains no package: The manifest is virtual ... ", ws.root()), // the old error message, to avoid fixing the tests and to special-case presumably the most common errro
  _ => bail!("no packages to compile"),
};
```

Let's also remove the `else`: the "then" branch of `if` is diverging, so we don't really need that else.

Reply:
Oh ok, ignore that commit I just made then, I'll revert it. hmmm I wasn't able to get that match statement to work. is the `Packages::All` branch guaranteed to run only with virtual manifests? I tried a match on `Workspace::is_virtual` and that fixed both tests now.